sentry_sdk/integrations/django/middleware.py

```diff
@@ -128,11 +134,27 @@ def __getattr__(self, method_name):
 
         def __call__(self, *args, **kwargs):
             # type: (*Any, **Any) -> Any
+            if self.async_capable:
+                return self.__acall__(*args, **kwargs)
+
             f = self._call_method
             if f is None:
                 self._call_method = f = _get_wrapped_method(self._inner.__call__)
             return f(*args, **kwargs)
 
+        async def __acall__(self, *args, **kwargs):
```

Comment:
There are two problems with this approach:

* we need to support Python 2, so this will be invalid syntax
* the wrapped method will not have correct span timings (which might be fine just to fix your issue)

We usually fix this by splitting out everything with async syntax into separate files that we import conditionally. let me know if that's something you want to work on, otherwise I can take over this PR

Reply:
@untitaker Ahh, I didn't think about python 2 support here :sweat_smile: 

Okay, I can make the files splitting and conditional import later today :slightly_smiling_face:  

> the wrapped method will not have correct span timings (which might be fine just to fix your issue)

I'm not really sure what you mean by this :thinking: Is that an issue with the error reporting ?

Reply:
the reason we wrap middlewares at all is to report middleware execution times for our [performance](https://sentry.io/for/performance/) product. if you wrap an async function with `_get_wrapped_method` you will get a wrapper that starts the timer, calls the async function but then stops the timer. Instead it should await the async function before stopping the timer.

Basically if we have async middlewares with this patch we will report 0ms timings, because the wrapped middleware immediately returns a future/coroutine.

Reply:
Ahh, I see what you mean. 
Okay, I'll jump into the code later and will try to find some solution to this :thinking: 
Not really sure how to test it though

Reply:
@untitaker I just make another iteration here.  
I haven't changed anything related to the span timings. I'm not sure it's actually possible (or if worth) to do this here since the code inside is supposed to run asynchronously :thinking:

Reply:
Just pushed a quick commit to (hopefully) fix the build

Reply:
@untitaker One more thing: I found that part of the tests are being skipped in pytest conditionaly.  (the ones that are failing). 
How can I run them? I can't find the info in https://github.com/getsentry/sentry-python/blob/master/CONTRIBUTING.md :slightly_smiling_face:

Reply:
The text buries the lede but you can control it through the makefile. `make help` is the actual entrypoint for test running.

Reply:
![image](https://user-images.githubusercontent.com/12683214/100880379-51bd0880-34b5-11eb-981b-a872a00dc235.png)
something is not quite right with the makefile I think :(

Reply:
I see, it's broken outside of macos. I fixed the bug in master. Can you try again?

Reply:
Still broker:/
```
./scripts/runtox.sh: 26: ./scripts/runtox.sh: Bad substitution
Makefile:36: recipe for target 'test-all' failed
make: *** [test-all] Error 2
```
I rebase to your last commit - `7dad958e`

Reply:
Ok, as a workaround: To run django tests you need to install pytest-django and django in your virtualenv manually.

Or you use `tox -e py3.7-django-2.2` as one example of many environments we test.

Reply:
Yes, it's working now, thanks

However I'm still struggling to make it working both for WSGI and ASGI. The patch inside `integrations/django/middleware.py` that dynamically changes BaseHandler's behaviour, makes it dificult to add asgi or wsgi specific behaviour://

I'll try again when I have more time

Reply:
I wonder why you need to do WSGI specific things? Wouldn't the middleware behave the same? I mean that's the sort of abstraction Django is supposed to provide, at least.

Reply:
@untitaker I don't need to. But every time I change something with the middleware, one of the handlers - wsgi or asgi breaks. If you look at the current state of branch, it's working fine with the ASGIHandler, but if you run `$python manage.py runserver --noasgi`, you'll see the error  "corouting object does not have attribute ..... " (or sth like that).

So I was trying to find a solution that won't introduce this regression. But the problem is that I cannot easily plug into and patch only the **ASGIHandler** since the sentry sdk is already patching the `load_middleware` which is part of the **BaseHandler**, so I cannot change the behaviour asgi-specifically. I tried different thing, but I couldn't manage to make them both working correctly (and I didn't have the time to continue debugging:/)